### PR TITLE
feature: support RetryStrategy for training jobs

### PR DESCRIPTION
**Pull request by spanghal03**
*Issue #, if available:*

*Description of changes:*
Today when a customer’s training job fails with an ISE, they have to manually retry the job using their custom retry strategy. With Resilient Training feature(RTJ), sagemaker will retry customers job on ISE’s.
- Adding max_retry_attempts in CreateTrainingJob and DescribeTrainingJob
-  Allow setting max_wait when max_retry_attempts  is set

*Testing done:*
- Added Unit and updated an Integ test
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/estimator.py [hunk 1]
```diff
@@ -1489,6 +1504,11 @@ def _get_train_args(cls, estimator, inputs, experiment_config):
         if estimator.enable_network_isolation():
             train_args["enable_network_isolation"] = True
 
+        if estimator.max_retry_attempts is not None:
+            train_args["retry_strategy"] = {"MaximumRetryAttempts": estimator.max_retry_attempts}
+        else:
+            train_args["retry_strategy"] = None
```
Comment (icywang86rui) [inline src/sagemaker/estimator.py:1510]: is it necessary to set this to None? Can we just leave it out

File: src/sagemaker/estimator.py [hunk 2]
```diff
@@ -1816,6 +1838,11 @@ def __init__(
                 will be disabled (default: ``False``).
             environment (dict[str, str]) : Environment variables to be set for
                 use during training job (default: ``None``)
+            max_retry_attempts (int): The number of times to move a job to the STARTING status.
+                You can specify between 1 and 30 attempts.
+                If the value of attempts is greater than one, the job is retried on InternalServerFailure the same number of attempts as the value.
+                You can cap the total duration for your job by setting ``max_wait`` and ``max_run``
+                (default: ``None``)
```
Comment (icywang86rui) [inline src/sagemaker/estimator.py:1845]: Please explain here that if this is set to None there will be no retries.

Comment (icywang86rui) [reply src/sagemaker/estimator.py:1845]: i am wondering if we should set the default to 3 or 2.

File: src/sagemaker/estimator.py (hunk 1 above)
Comment (spanghal03) [reply src/sagemaker/estimator.py:1510]: As per this test, if I have to assert that retry_strategy is None when max_retry_attempts are not set, then yes we have to explicitly set to None https://github.com/aws/sagemaker-python-sdk/blob/master/tests/unit/test_estimator.py#L1060 . given this my understanding is that we have to set it to None explicitly

File: src/sagemaker/estimator.py (hunk 2 above)
Comment (spanghal03) [reply src/sagemaker/estimator.py:1845]: We can set the default to 3 or 2 as we want to customers to explicitly opt-in. However I think I need to re-word it like " If the value of attempts is greater than zero, the job is retried on InternalServerFailure the same number of attempts as the value.
                 You can cap the total duration for your job by setting ``max_wait`` and ``max_run``"

Comment (icywang86rui) [reply src/sagemaker/estimator.py:1845]: yeah, that is more clear. Let's just do None for the default for now.

File: src/sagemaker/estimator.py (hunk 1 above)
Comment (icywang86rui) [reply src/sagemaker/estimator.py:1510]: hmm, k

File: src/sagemaker/estimator.py (hunk 2 above)
Comment (spanghal03) [reply src/sagemaker/estimator.py:1847]: done

**Review by icywang86rui — APPROVED** [review]

**Pull request merged**